spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverter.java

```diff
@@ -0,0 +1,281 @@
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
+
+		persistentEntity.doWithProperties(
+				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+					Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+					accessor.setProperty(datastorePersistentProperty, value);
+				});
+
+		return instance;
+	}
+
+	@Override
+	public void write(Object source, Entity.Builder sink) {
+		DatastorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(source.getClass());
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(source);
+		persistentEntity.doWithProperties(
+				(DatastorePersistentProperty datastorePersistentProperty) ->
+						writeProperty(sink, accessor, datastorePersistentProperty));
+	}
+
+	/**
+	 * In order to support {@link CustomConversions}, this method applies 2-step conversion.
+	 * The first step produces one of {@link SimpleTypeHolder}'s simple types.
+	 * The second step converts simple types to Datastore-native types.
+	 * The second step is skipped if the first one produces a Datastore-native type.
+	 *
+	 * @param sink entity builder where the value should be written to
+	 * @param accessor property accessor
+	 * @param persistentProperty persistent property
+	 */
+	private void writeProperty(Entity.Builder sink, PersistentPropertyAccessor accessor,
+			DatastorePersistentProperty persistentProperty) {
+		Object propertyVal = accessor.getProperty(persistentProperty);
+		if (propertyVal != null) {
+			TwoStepConversion twoStepConversion = new TwoStepConversion(
+					propertyVal.getClass(),
+					this.customConversions,
+					this.toNativeConversions);
+			if (twoStepConversion.getFirstStepTarget() != null) {
+				propertyVal = this.conversionService.convert(propertyVal, twoStepConversion.getFirstStepTarget());
+			}
+
+			if (twoStepConversion.getSecondStepTarget() != null) {
+				propertyVal = this.conversionService.convert(propertyVal, twoStepConversion.getSecondStepTarget());
+			}
+		}
+		Value val = getDatastoreWrappedValue(propertyVal, persistentProperty);
+		sink.set(persistentProperty.getFieldName(), val);
+	}
+
+	@SuppressWarnings("unchecked")
+	private Value getDatastoreWrappedValue(Object propertyVal, DatastorePersistentProperty persistentProperty) {
+		if (propertyVal == null) {
+			return new NullValue();
+		}
+		Function wrapper = DATASTORE_TYPE_WRAPPERS.get(propertyVal.getClass());
+		if (wrapper != null) {
+			return (Value) wrapper.apply(propertyVal);
+		}
+		throw new DatastoreDataException("Unable to convert a property" +
+				" with name " + persistentProperty.getFieldName()
+				+ " to Datastore supported type. The property's type is " + propertyVal.getClass());
+	}
+
+	static class TwoStepConversion {
+		private Class<?> firstStepTarget;
+
+		private Class<?> secondStepTarget;
+
+		TwoStepConversion(Class<?> firstStepSource, CustomConversions customConversions,
```

Comment:
Since you're doing all this work in the constructor, maybe just create a private method instead, and return a `TwoStepConversion` bean.